src/betterproto/__init__.py

```diff
@@ -519,23 +519,67 @@ def __post_init__(self) -> None:
             if meta.group:
                 group_current.setdefault(meta.group)
 
-            if getattr(self, field_name) != PLACEHOLDER:
-                # Skip anything not set to the sentinel value
+            if self.__raw_get(field_name) != PLACEHOLDER:
+                # Found a non-sentinel value
                 all_sentinel = False
 
                 if meta.group:
                     # This was set, so make it the selected value of the one-of.
                     group_current[meta.group] = field_name
 
-                continue
-
-            setattr(self, field_name, self._get_field_default(field_name))
-
         # Now that all the defaults are set, reset it!
         self.__dict__["_serialized_on_wire"] = not all_sentinel
         self.__dict__["_unknown_fields"] = b""
         self.__dict__["_group_current"] = group_current
 
+    def __raw_get(self, name: str) -> Any:
+        return super().__getattribute__(name)
+
+    def __eq__(self, other):
+        if type(self) != type(other):
+            return False
+
+        equal = True
+        for field_name in self._betterproto.meta_by_field_name:
+            self_val = self.__raw_get(field_name)
+            other_val = other.__raw_get(field_name)
+            if self_val is PLACEHOLDER and other_val is PLACEHOLDER:
+                continue
+            elif self_val is PLACEHOLDER:
+                self_val = self._get_field_default(field_name)
+            elif other_val is PLACEHOLDER:
+                other_val = other._get_field_default(field_name)
+
+            if self_val != other_val:
+                equal = False
+                break
+
+        return equal
+
+    def __repr__(self):
+        parts = [self.__class__.__name__, "("]
+        found = False
+        for field_name in self._betterproto.meta_by_field_name:
+            value = self.__raw_get(field_name)
+            if value is PLACEHOLDER:
+                continue
+            found = True
+            parts.extend([field_name, "=", repr(value), ","])
```

Comment:
There should be a space after this comma, to match PEP8

Reply:
I think it would be better to join them using ", " to avoid this.

On another note, since this is for constructing the class from the repr you should probably also check if the field is an Enum and then use `Enum.__name__.value.name`

Reply:
I'm not sure I follow this:

> On another note, since this is for constructing the class from the repr you should probably also check if the field is an Enum and then use Enum.__name__.value.name

`repr` isn't generally meant to be round-trippable.  For example:

```python
from dataclasses import dataclass
from enum import Enum

class Color(Enum):
    RED = 1
    BLUE = 2

@dataclass
class Foo(object):
    color: Color

repr(Foo(Color.RED)) # => "Foo(color=<Color.RED: 1>)"
```

This implementation of `__repr__` gives a similar result to what the implementation provided by `@dataclass` gives.

Reply:
According to the python docs (https://docs.python.org/3/reference/datamodel.html#object.__repr__) where possible an object's `__repr__` should allow for `x = eval(repr(x))` where possible, with the current implementation it would not be possible to do this, so I was suggesting that Enum fields should be converted to there Enum class name and member name to allow for this.

Reply:
I agree with your interpretation of the docs, but already the built in `enum.Enum` does not follow this convention.  I'm just delegating to each field's existing `__repr__` implementation, which is as much as we should do, in my opinion.  I don't think `Message`'s implementation of `__repr__` should inspect its field's types and change it's behavior.

Reply:
I just realized that `betterproto` generates enums that inherit from `betterproto.Enum`, not `enum.Enum`.  I apologize for my confusion.

Given that, it would be possible to implement `betterproto.Enum.__repr__` and get the behavior you're talking about.  And it seems like a perfectly good thing to do.  It's independent of this PR, though, and would have the same benefit whether it lands before or after this, so I still wouldn't want to tie the two changes together.

Reply:
also in favor of keeping general improvements out of scope of bugfix PR's, unless they are essential

Reply:
I wasn't suggesting altering the `betterproto.Enum.__repr__` but altering `Message.__repr__` to use `enum.Enum.__str__` instead to follow the standard.